### ISSUE TITLE
Ability to query remote named graphs

### ISSUE THREAD

**Issue by frmichel** [w3c/sparql-12#18]
## Why?

Clauses referring to named graphs assume these are stored in the local triple store. E.g.
```
SELECT FROM <mygraph> WHERE {   }
SELECT ... WHERE { GRAPH <mygraph> { ... } }
```

But this does not allow to query a named graph from a remote endpoint, in short a "remote named graph".

## How could it look like ?

An endpoint could evaluate a "FROM <graph>" or "GRAPH <graph> { }" by first looking for the named graph in the local triple store, then, if it cannot be found, try to dereference its URI and fetch its content locally.

## Previous work

None that I'm aware of. But this may be somehow related to issue #15 about a "LOAD WHERE" clause.

## Considerations for backward-compatibility

Possibly conflicting: with SPARQL 1.1, if a named graph does not exist in the local triple store, then what happens: is it created or just considered empty?

**Comment by pietercolpaert** [comment]
I think this is very close to what is suggested in https://github.com/w3c/sparql-12/issues/10, but with the SERVICE clause?

**Comment by cygri** [comment]
> evaluate by first looking for the named graph in the local triple store, then, if it cannot be found, try to dereference its URI and fetch its content locally.

While I agree with the general requirement, this specific part is a terrible idea. We know this at TopQuadrant because we've implemented pretty much what you describe, and had to removed it again because it caused so many problems.

Imagine a system that is supposed to only use local graphs, but uses non-dereferenceable HTTP IRIs to name them. If an expected graph is missing for whatever reason, the SPARQL processor would try to fetch it from the web, leading to a slow HTTP request and bandwidth costs. In a busy back-end system, this could have disastrous consequences.

Special syntax is needed for the case where the query goes out to the web.

**Comment by cygri** [comment]
Syntax straw man:

```
SELECT ... FROM LOCAL <myGraph> WHERE { ... }
SELECT ... FROM REMOTE <myGraph> WHERE { ... }
... LOCAL GRAPH <myGraph> { ... } ...
... REMOTE GRAPH <myGraph> { ... } ...
```
So, whenever using `FROM` or `GRAPH`, one can add `LOCAL` or `REMOTE` to explicitly specify whether the graph IRI is to be interpreted as a named graph in the service's dataset, or as a URL on the web. If neither `FROM` nor `LOCAL` is given, we stick with SPARQL 1.1 semantics (that is, it's entirely up to the processor what the graph IRI refers to.)

**Comment by frmichel** [comment]
@cygri 
> whenever using FROM or GRAPH, one can add LOCAL or REMOTE to explicitly specify whether the graph IRI is to be interpreted as a named graph in the service's dataset, or as a URL on the web. If neither FROM nor LOCAL is given, we stick with SPARQL 1.1 semantics

Yes you are probably right, and I trust your experience in this respect.

Also, because the recommendation specifies a feature does not mean it must always be supported. I would imagine this feature as possibly deactivated by configuration, pretty much like public endpoints usually do not SERVICE clauses to remote domains.

**Comment by frmichel** [comment]
@pietercolpaert : it could be seen as an extension of the SERVICE clause, as suggested in #10, yet the ability to query named graphs is a bit different.

A named graph is not necessarily dereferenceable, and if it is, the result is not necessarily the content of the graph, it can be a few triples giving metadata about the named graph.
Therefore, a `SERVICE <named graph URI>` would not be guaranteed to return anything nor to return what you would expect.
